### DISCUSSION TITLE
MPE - MIDI Polyphonic Expression

### DISCUSSION THREAD

**Discussion by tobia** [probonopd/MiniDexed#591] [Ideas]
Hello. Thank you to probono and to everybody else who is contributing to this amazing project! I'm in the process of acquiring suitable hardware to run MiniDexed and maybe start contributing to it too.

I've always been interested in FM synthesis, but I'm also an alternative MIDI controller enthusiast, so I was wondering if the current version supports MPE.

For those not in the know, MPE or MIDI Polyphonic Expression is a method of using MIDI messages that enables pitch bend, mod wheel, and other CC parameters to be adjusted continuously for individual notes in a chord. It was accepted in 2017-2018 by the MIDI associations, so it's now part of the MIDI standard, and it's supported by an increasing number of instruments (Continuum Fingerboard, LinnStrument, ROLI Seaboard, Sensel Morph, Eigenharp, Erae Touch...) 

Technically, ***MPE repurposes the 16 MIDI channels as notes instead of instruments.*** This is a tradeoff (or a hack) as it limits you to only have a single instrument on each port or cable (which is usually the case anyways) and it limits you to a maximum of 16‑polyphony, one note per channel (or alternatively 2 voices with 8-polyphony each, when you are using a split keyboard). In exchange for that, it lets you control pitch‑bend, mod wheel and other CC independently for each note, meaning you can emulate single‑string vibrato and other expressions traditionally available on acoustic instruments, if you have an appropriate controller, without having the vibrato applied across the whole instrument.

From the outside, the frontend part doesn't seem too hard to implement, it's just a couple of system menu options to select the preferred MPE mode, and a few if statements to route the messages to the right places. But then, since the whole point is to be able to control pitch-bend, mod-wheel, after-touch, filter cutoff, and any other available MIDI parameters independently for each note, it's dependent on whether the synth engine is prepared to receive these messages separately for each note.

Does anybody know what the current capabilities of MiniDexed are on this regard?

**Comment by probonopd** [comment]
Is this the same as https://github.com/probonopd/MiniDexed/issues/572?

**Reply by tobia** [reply]
I think #572 is just part of it. Basically, when you set a MiniDexed voice to MPE mode, it would listen to several channels for note events (usually all 16, but sometimes the first 8 or the last 8) and then, crucially, when it receives any control parameter, including:

1. pitch bend
2. mod wheel
3. after touch
4. and any other CC

on one of those channels, it would only affect the note that was received on that channel.

The purpose is, if you have a MIDI controller that allows you to move your fingers freely on the surface, you can "slide" each note of a chord independently of the others, and the controller would send pitch bend events on that note's channel, and the pitch bend would not influence the other notes that are being played with other fingers, that were send on different channels.

**Reply by BobanSpasic** [reply]
@tobia - does the MIDI controller send on different channels, or it sends on one channel and the sound module needs to distribute the notes to the separate channels?
If it sends on different channels, you already have 8 Tone Generators with MiniDexed, every one of them with its own MIDI channel. Every Tone Generator has its own processing of pitch bend, MW etc.
As for the other CCs, take a look at the Wiki: https://github.com/probonopd/MiniDexed/wiki/MIDI-Implementation

**Comment by diyelectromusic** [comment]
The short answer is no.

The long answer is that by supporting eight tone generators and effectively running the device as 8 mono-synths each on their on MIDI channel, it could, kind of, be simulated at the DAW end, but the hardware doesn't do anything special to support it.

It is the kind of thing that could be simulated in MiniDexed by a different polyphonic note-selection algorithm before actually calling out to the Synth_Dexed tone engines.  It might be possible to wind up the number of tone generators and drop down the polyphony of each for example.,  We would either have to decide on a fixed "zone" or implement the suggested MIDI Zone control messages.

But it would be quite a fudge really and quite far away from the original DX7.  It would also only work at the voices level and not with performances.  I'm pretty sure the Synth_Dexed engine itself doesn't support MPE.

Whilst there is an adopted MPE specification that I think is aimed for MIDI 1.0 devices to support, proper MPE (I believe) really only comes in with MIDI 2.0 which has extra properties on channel voice messages and iirc some new channel voice messages too for such things.  But MIDI 2.0 is a lot more complex than MIDI 1.0...

Kevin

**Reply by diyelectromusic** [reply]
It also is unlikely to be compatible with serial MIDI.  See this example in the spec:

![image](https://github.com/probonopd/MiniDexed/assets/68612569/f75a09c9-7f10-4db6-91dc-645d8fa7ff4b)

Each NoteOn goes from three bytes to potentially 11.  That is quite a performance hit and also means that no MIDI Running Status is possible (where NoteOn/NoteOff messages are optimised down to 2 bytes each).

Kevin

**Reply by tobia** [reply]
I plain MPE is just MIDI 1.0 where each channel is used for a different note. Yes, I guess using each of the 8 synths on a different channel would be a basic way to make it work.

**Reply by probonopd** [reply]
Does MPE in Dexed VSTi3 work for you?

**Reply by silentbirdie** [reply]
@probonopd i will test out dexed mpe support and come back with my findings.

**Reply by silentbirdie** [reply]
@probonopd so far in my testing it is supporting poly pitchbend with my exquis mpe controller with the x axis and nothing else but that is probably a patch problem I feel. I need to figure out how to make an mpe patch for the synth.

**Comment by vorptronica** [comment]
I'm very curious about the state of miniDexed mpe too. I just built my own targeting my Linnstrument. I'm planning to embed it into the side rail (custom 3d print) so my Linnstrument always can make some synth sounds without having to attach it to anything.
I haven't tried dexed yet but I'll grab it and see how it performs. Even if it does pitchbend it would be great to port that over to MiniDexed. Is that a challenging prospect?

**Comment by probonopd** [comment]
Hi @vorptronica. Currently I don't have a compatible hardware to test with. What should MPE modulate? The same as Pitch Bend? Or should it be possible to set this up separately, similar to how you can set it up for Pitch Bend, Breath Controller, and Pedal? This poses an interesting question, as the voices ("patches") made for the DX7 don't know MPE.

**Reply by vorptronica** [reply]
Thanks for the reply probonopd. Pitch bend would be the main one, but the other "axes" of a Linnstrument are after touch (pressure on each pad after initial note activation) and vertical finger movement. For each pad event the synth takes the next available channel (often the algorithm takes the channel with the longest no-signal interval...ideally not cutting off a previous note). All the pitch bend, after touch, and timbre change messages are then sent only only on that channel. They are sent as the normal messages for a non-mpe controller, IIRC. The synth then reads the channel and identifies the correct voice to get the modulation from those CC changes. This allows each pad to bend notes in any direction independently, with the restriction that you have a limit of 16 note polyphony. All the signals conform to the standard midi ones so the main programming effort is to choose the next available voice and limit the bend, after touch, and timbre changes to the chosen one. Apologies if that was already all known to you. Understandably this is exiting a pure DX7 emulator scenario but I think it would be fantastic to have for my project. As an option to enable for those of us that need it would be very cool. Presumably this logic was all added to dexed by the Surge team if I read those changelist notes correctly. I still need to test that myself.
